### PR TITLE
Ui grid refactoring and row filter.

### PR DESCRIPTION
**Pull request by PiotrZakrzewski**
Refactored cohort-grid with angular-ui-grid. 
Added row filters for each column. 
Issue: I could not write unit-tests for it, I do not know how to properly mock firing a broadcast. Any suggestions appreciated.

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -11,17 +11,17 @@
     "angular-ui-router": "~0.2.15",
     "bootstrap": "~3.3.6",
     "angular-bootstrap": "~1.3.2",
-    "angular": "~1.5.0",
+    "angular": "~1.5.5",
```
Comment (PiotrZakrzewski) [inline bower.json]: @rnugraha Actually I did not intend to make this change - do you think we should upgrade to 1.5.5 ? If not I will remove it.

File: src/app/components/cohort-grid/cohort-grid.less [hunk 2]
```diff
@@ -1,3 +1,7 @@
 .ts-export-btn {
   float: right;
 }
+.ts-cohort-grid {
+  width: 600px;
```
Comment (PiotrZakrzewski) [inline src/app/components/cohort-grid/cohort-grid.less]: @rnugraha ui-grid does not seem to play nice with width set in %

File: bower.json (hunk 1 above)
Comment (rnugraha) [inline bower.json]: if it's compiled then should be fine

File: src/app/components/cohort-grid/cohort-grid.less (hunk 2 above)
Comment (rnugraha) [inline src/app/components/cohort-grid/cohort-grid.less]: just wondering is it possible to make the width is adjusted when a node is added? e.g. .. add first node width = 200px, add 2nd node width = 400px;

Comment (PiotrZakrzewski) [inline src/app/components/cohort-grid/cohort-grid.less]: fixed (see latests commit added to this PR)

File: src/app/components/cohort-grid/cohort-grid.directive.js [hunk 3]
```diff
@@ -3,36 +3,108 @@
...
+          enableFiltering: true,
+          onRegisterApi: function (gridApi) {
+            $scope.gridApi = gridApi;
+          }
+        };
+
+
+        $scope.$watch('cohort', function () { // I am leaving these two watchers so that we can find out where the
+                                              // issue is with cohorts reset
+          console.log('Watcher registered change in cohorts');
```
Comment (rnugraha) [inline src/app/components/cohort-grid/cohort-grid.directive.js]: pls remove

File: src/app/components/cohort-grid/cohort-grid.directive.js [hunk 4]
```diff
@@ -3,36 +3,108 @@
...
+        }, true);
+
+        $scope.$watch('headers', function () { // I am leaving these two watchers so that we can find out where the
+                                               // issue is with cohorts reset
+          console.log('Watcher registered change in headers');
```
Comment (rnugraha) [inline src/app/components/cohort-grid/cohort-grid.directive.js]: pls remove

File: src/app/components/cohort-grid/cohort-grid.directive.js [hunk 5]
```diff
@@ -3,36 +3,108 @@
...
+
+        $scope.$on('collectionUpdated', function (event, passedFromBroadcast) {
+          console.log('collectionUpdated event fired');
+          //TODO: refactor the directive to either use only data from event or from data binding. Not both.
+          var _cohorts = passedFromBroadcast[0];
+          var _headers = passedFromBroadcast[1];
+          $scope.gridOptions.data.length = 0;       // In order to make sure ui-grid is refreshed also when tables
+          $scope.gridOptions.columnDefs.length = 0; // are of the same length
+          $scope.style_modifiers.width = 0; // as above
+          $timeout(function () { // this is necessary for ui-grid to notice the change at all
```
Comment (rnugraha) [inline src/app/components/cohort-grid/cohort-grid.directive.js]: why not $watch gridOptions?

Comment (PiotrZakrzewski) [inline src/app/components/cohort-grid/cohort-grid.directive.js]: ng-grid has its own internal way of watching changes to it, limitation of their approach is that any change to the gridOptions need to be done in two steps to guarantee that that the changes will be applied to the displayed table: first set length value of gridOption fields to zero and then in next digest set assign given field with desired value. 
See http://stackoverflow.com/questions/37485045/angular-ui-grid-does-not-show-content-unless-browser-window-is-resized for reference

**Pull request merged**